[wasm][debugger] View object attributes - DebugType=full

thaystg opened the pull request:
Fix view object attributes from a class defined in an assembly with debugType = full.

Fix discussion on: https://github.com/mono/mono/issues/19942

Adding the Runtime.consoleAPICalled, I couldn't use the Log.entryAdded didn't work on VS.

in VS:
![image](https://user-images.githubusercontent.com/4503299/144918534-d1b89a42-90a6-43ca-8ea7-1de6d561b325.png)

in Chrome:
![image](https://user-images.githubusercontent.com/4503299/144920544-62381330-2750-470d-a38b-b45d170183ae.png)

thaystg committed:
Fix view variable values from an assembly with debugType = full

ilonatommy commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:
```diff
@@ -849,3 +849,16 @@ public static void RunDebuggerBreak()
+        var myType = asm.GetType("DebuggerTests.ClassToInspectWithDebugTypeFull");
+        var myMethod = myType.GetConstructor(new Type[] { });
+        var a = myMethod.Invoke(new object[]{});
+        Console.WriteLine(a);
```
Should we leave this writing to the console here?

thaystg replied:
Removed :)

thaystg committed:
Addressing @ilonatommy comment.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
+                }
+                catch (BadImageFormatException)
+                {
+                    Console.WriteLine($"Warning: Unable to read debug information of: {Name} (use DebugType=Portable/Embedded)");
```
you could log this using https://chromedevtools.github.io/devtools-protocol/tot/Log/ have it surface to the frontend

lewing replied:
You emit https://chromedevtools.github.io/devtools-protocol/tot/Log/#event-entryAdded from the proxy?

thaystg replied:
As far as I understood this is not supported by JSDebugger so I implemented using `Runtime.consoleAPICalled`.

lewing commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj:
```diff
@@ -22,6 +22,7 @@
     <ProjectReference Include="..\library-dependency-debugger-test2\library-dependency-debugger-test2.csproj" Private="true"/>
     <ProjectReference Include="..\debugger-test-with-source-link\debugger-test-with-source-link.csproj" Private="true"/>
     <ProjectReference Include="..\ApplyUpdateReferencedAssembly\ApplyUpdateReferencedAssembly.csproj" />
+    <ProjectReference Include="..\debugger-test-with-full-debug-type\debugger-test-with-full-debug-type.csproj" Private="true"/>
```
does this need to be added?

thaystg replied:
I think so, to compile the new project for the test case.

lewing replied:
right, I mean did you forget to add the project to the pr?

thaystg replied:
I'm sorry Larry!

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
-                pdbMetadataReader = MetadataReaderProvider.FromPortablePdbStream(pdbStream).GetMetadataReader();
+                try
+                {
+                    pdbMetadataReader = MetadataReaderProvider.FromPortablePdbStream(pdbStream).GetMetadataReader();
```
which method throws `BadImageFormatException`?

thaystg replied:
`FromPortablePdbStream`

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
+            AssemblyNameUnqualified = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
             if (pdb != null)
             {
                 pdbStream = new MemoryStream(pdb);
```
IIUC, `pdbStream` doesn't really ever get used again. We instantiate it here, and use it to get the `pdbMetadataReader`. And in other methods, we re-instantiate.
And same for `asmReader`.
If so, then we can remove the fields, and dispose these.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -541,10 +541,19 @@ public unsafe AssemblyInfo(string url, byte[] assembly, byte[] pdb)
             asmStream = new MemoryStream(assembly);
             peReader = new PEReader(asmStream);
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            Name = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
+            AssemblyNameUnqualified = asmMetadataReader.GetAssemblyDefinition().GetAssemblyName().Name + ".dll";
```
Looking at the implementation, `GetAssemblyDefinition()` returns a new object on every call. And `GetAssemblyName` constructing a new `AssemblyName` instance. so, we should avoid doing that twice.

thaystg committed:
Adding the csproj for the new test.

thaystg committed:
Adressing @radical comments.

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Removing method duplicated.
Keep peReader as an attribute as is recommended in the documentation. https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.pereaderextensions.getmetadatareader?view=net-6.0#System_Reflection_Metadata_PEReaderExtensions_GetMetadataReader_System_Reflection_PortableExecutable_PEReader_

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -526,37 +526,49 @@ internal class AssemblyInfo
+            PdbGuid = codeViewData.Guid;
+            PdbName = codeViewData.Path;
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            var asmDef = asmMetadataReader.GetAssemblyDefinition();
```
removed the `AssemblyNameUnqualified `as it has the same value of `Name`

thaystg committed:
Calling Runtime.consoleAPICalled.

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -526,37 +526,49 @@ internal class AssemblyInfo
+            var asmStream = new MemoryStream(assembly);
+            var peReader = new PEReader(asmStream);
+            var entries = peReader.ReadDebugDirectory();
+            var codeView = entries[0];
```
done.

thaystg commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -526,37 +526,49 @@ internal class AssemblyInfo
             asmMetadataReader = PEReaderExtensions.GetMetadataReader(peReader);
+            var asmDef = asmMetadataReader.GetAssemblyDefinition();
+            Name = asmDef.GetAssemblyName().Name + ".dll";
+            AssemblyNameUnqualified = asmDef.GetAssemblyName().Name + ".dll";
```
removed this property. as far as I undesrtood it has the same content of the `Name` property.

thaystg committed:
addressing @radical comment.
